### PR TITLE
Include a function for paying block relayers

### PR DESCRIPTION
**Pull request by lrubiorod**

### CODE REVIEW

**Review by mariocao — CHANGES_REQUESTED** [review]
Fix format warnings :)

File: contracts/ActiveBridgeSetBlockRelay.sol [hunk 1]
```diff
@@ -268,6 +268,19 @@ contract ActiveBridgeSetBlockRelay is BlockRelayInterface {
     return blocks[_blockHash].relayerAddress;
   }
 
+  /// @dev Pay the block reward to the relayer in case it has not been paid before
+  /// @param _blockHash Hash of the block header
+  function payRelayer(uint256 _blockHash) external payable override {
```
Comment (clbartoli) [inline contracts/ActiveBridgeSetBlockRelay.sol:273]:
```suggestion
  function payRelayer(uint256) external payable override {
```

Comment (clbartoli) [reply contracts/ActiveBridgeSetBlockRelay.sol:273]: this change should remove the warnings

File: test/cent_block_relay.js [hunk 2]
```diff
@@ -36,6 +45,63 @@ contract("Centralized Block relay", accounts => {
...
+            value: ether("0"),
+            gasPrice: 1,
+          }
+        ),
+        "Block reward should cover gas expenses. Check the estimateGasCost method")
+    })
+
+    it("should pay to the block relayer", async () => {
+      const expectedId = "0x" + sha.sha256("first id")
+      // First, we use the call to check the boolean result
```
Comment (clbartoli) [inline test/cent_block_relay.js:63]:
```suggestion
      // First, we use the check the boolean result
```

File: test/cent_block_relay.js [hunk 3]
```diff
@@ -36,6 +45,63 @@ contract("Centralized Block relay", accounts => {
...
+      const result = await blockRelayInstance.isRelayerPaid.call(expectedId)
+      assert.equal(result, false)
+
+      // Later we use the real value transfer
```
Comment (clbartoli) [inline test/cent_block_relay.js:67]:
```suggestion
      // Later we transfer the reward
```

File: contracts/ActiveBridgeSetBlockRelay.sol (hunk 1 above)
Comment (lrubiorod) [reply contracts/ActiveBridgeSetBlockRelay.sol:273]: image

File: contracts/ActiveBridgeSetBlockRelay.sol [hunk 4]
```diff
@@ -49,6 +51,9 @@ contract ActiveBridgeSetBlockRelay is BlockRelayInterface {
     Hashes voteHashes;
   }
 
+  // Block reporting is not subject to increases
```
Comment (mariocao) [inline contracts/ActiveBridgeSetBlockRelay.sol:54]:
```suggestion
  // Maximum amount of gas for reporting a Block (not subject to increases)
```

File: contracts/ActiveBridgeSetBlockRelay.sol [hunk 5]
```diff
@@ -268,6 +273,19 @@ contract ActiveBridgeSetBlockRelay is BlockRelayInterface {
     return blocks[_blockHash].relayerAddress;
   }
 
+  /// @dev Pay the block reward to the relayer in case it has not been paid before
+  /// @param _blockHash Hash of the block header
+  function payRelayer(uint256 _blockHash) external payable override {
+    // TODO Review this function in that kind of bridge
```
Comment (mariocao) [inline contracts/ActiveBridgeSetBlockRelay.sol:279]: Not implemented yet for this bridge implementation

File: contracts/CentralizedBlockRelay.sol [hunk 6]
```diff
@@ -239,5 +244,37 @@ contract CentralizedBlockRelay is BlockRelayInterface {
...
+  function estimateGasCost(uint256 _gasPrice) public pure returns(uint256){
+    return SafeMath.mul(_gasPrice, MAX_REPORT_BLOCK_GAS);
+  }
+
+  /// @dev Ensures that rewards cover the cost of post a block in the Block Relay
+  /// @param _blockReward The amount for rewarding the reporting of the blocks
+  function isPayingGasCosts(uint256 _blockReward) internal view {
+    uint256 minBlockReward =  estimateGasCost(tx.gasprice);
+    require(_blockReward >= minBlockReward, "Block reward should cover gas expenses. Check the estimateGasCost method.");
+  }
```
Comment (mariocao) [inline contracts/CentralizedBlockRelay.sol:278]: This function could define as modifier

File: contracts/CentralizedBlockRelay.sol [hunk 7]
```diff
@@ -239,5 +244,37 @@ contract CentralizedBlockRelay is BlockRelayInterface {
...
+  function isRelayerPaid(uint256 _blockHash) public view override returns(bool){
+    return blocks[_blockHash].isPaid;
+  }
+
+  /// @dev Pay the block reward to the relayer in case it has not been paid before
+  /// @param _blockHash Hash of the block header
+  function payRelayer(uint256 _blockHash) external payable override {
+    require(isRelayerPaid(_blockHash) == false, "The relayer has already been paid");
+    // Check if rewards are covering gas costs
+    isPayingGasCosts(msg.value);
```
Comment (mariocao) [inline contracts/CentralizedBlockRelay.sol:259]: Use as modifier

File: contracts/BlockRelayProxy.sol [hunk 8]
```diff
@@ -112,6 +112,21 @@ contract BlockRelayProxy {
...
   }
 
+  /// @dev Pay the block reward to the relayer in case it has not been paid before
+  /// @param _blockHash Hash of the block header
+  function payRelayer(uint256 _blockHash, uint256 _epoch) external payable {
+    address controller = getController(_epoch);
+    return BlockRelayInterface(controller).payRelayer{value: msg.value}(_blockHash);
+  }
+
+  /// @dev This function checks if the relayer has been paid
```
Comment (mariocao) [inline contracts/BlockRelayProxy.sol:122]:
```suggestion
  /// @dev Checks if the block relayer has been paid
```

File: contracts/CentralizedBlockRelay.sol [hunk 9]
```diff
@@ -239,5 +251,28 @@ contract CentralizedBlockRelay is BlockRelayInterface {
     }
     return _root == tree;
   }
+  /// @dev This function checks if the relayer has been paid
```
Comment (mariocao) [inline contracts/CentralizedBlockRelay.sol:254]:
```suggestion

  /// @dev This function checks if the relayer has been paid
```

**Review by mariocao — APPROVED** [review]

**Pull request merged**